### PR TITLE
End of Rains - Evaporation 

### PR DESCRIPTION
**Pull request by bethanclarke**
R-code implemented for evaporation receiver, and linked logical and calculated columns to evaporation

### CODE REVIEW

**Comment by bethanclarke** [comment]
@shadrackkibet this is ready for review please. I think the evaporation label isn't thinking to the End of the Season check box and please test evaporation and missing values options in particular

**Comment by shadrackkibet** [comment]
Reviewing and testing this now.

**Comment by shadrackkibet** [comment]
This is working, I have been testing and i will continue to test it. I have got a number  issues to report;

- Labels for the evaporation are not linked up well - **Evaporation** &  **Replace NA with**- I suppose you know this.
- The evaporation receiver is not getting focus when variable radio button is checked.
- I think it should be possible to get evaporation column from a linked data frame.
- There is still brackets on Day Range filter expression- This should be removed so that it is consistent with other climatic dialogs.
- When end_of rains checkbox is checked and end of rains column name is empty ok should be disabled. and also, when Replace NA with input is empty?

File: instat/dlgEndOfRainsSeason.vb [hunk 1]
```diff
@@ -26,6 +30,13 @@ Public Class dlgEndOfRainsSeason
     Private clsWBWaterFilterMaxOperator, clsWBFirstWaterBalanceMaxOperator, clsPMaxOperatorMax As New ROperator
     '0
     Private clsWBReplaceNAMin, clsWBReplaceNAMinFunction, clsWBReplaceNAMinFunctionList, clsWBWaterBalanceMin, clsWBWaterFilterMin, clsWBWaterFilterMinList, clsWBFirstWaterBalanceMin, clsWBFirstWaterBalanceMinList, clsWBWaterBalanceMinList, clsReduceFunctionMin, clsPMinFunctionMin, clsPMaxFunctionMin As New RFunction
+    Private clsEvaporationReplaceNAFunc, clsEvaporationReplaceNA, clsEvaporationReplaceNAFuncList As New RFunction
+    Private clsReduceOpEvapMin, clsReduceOpEvapMax As New ROperator
+
+    Private Sub ucrChkEvaporationAsReceiver_Load(sender As Object, e As EventArgs)
```
Comment (shadrackkibet) [inline instat/dlgEndOfRainsSeason.vb]: do you need this sub?

File: instat/dlgMakeDate.Designer.vb [hunk 2]
```diff
@@ -41,24 +41,43 @@ Partial Class dlgMakeDate
         Me.components = New System.ComponentModel.Container()
         Dim resources As System.ComponentModel.ComponentResourceManager = New System.ComponentModel.ComponentResourceManager(GetType(dlgMakeDate))
         Me.grpTwoColumns = New System.Windows.Forms.GroupBox()
+        Me.ucrChkTwoDigitYear = New instat.ucrCheck()
```
Comment (shadrackkibet) [inline instat/dlgMakeDate.Designer.vb]: do you need makedate  files here?

File: instat/dlgEndOfRainsSeason.vb [hunk 3]
```diff
@@ -603,7 +630,52 @@ Public Class dlgEndOfRainsSeason
...
+    Private Sub CoreControls_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrChkEndOfRains.ControlContentsChanged, ucrChkEndOfSeason.ControlContentsChanged, ucrReceiverRainfall.ControlContentsChanged, ucrReceiverDate.ControlContentsChanged, ucrReceiverYear.ControlContentsChanged, ucrReceiverDOY.ControlContentsChanged, ucrNudCapacity.ControlContentsChanged, ucrNudWBLessThan.ControlContentsChanged, ucrInputWBColName.ControlContentsChanged, ucrNudTotalOverDays.ControlContentsChanged, ucrNudAmount.ControlContentsChanged, ucrReceiverEvaporation.ControlContentsChanged, ucrInputEvaporation.ControlContentsChanged, ucrInputReplaceNA.ControlContentsChanged
         TestOKEnabled()
+        Evaporation()
```
Comment (shadrackkibet) [inline instat/dlgEndOfRainsSeason.vb]: why do you have this sub here?

File: instat/dlgEndOfRainsSeason.vb [hunk 4]
```diff
@@ -14,10 +14,14 @@
 ' You should have received a copy of the GNU General Public License 
 ' along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+Imports instat
```
[the remark this answers is not in the document]
Comment (bethanclarke) [reply instat/dlgEndOfRainsSeason.vb]: I think we should not worry about these sorts of things right now, as it does not effect how the dialog works

**Comment by bethanclarke** [comment]
@shadrackkibet this is ready to test again. I have fixed these issues you pointed out: 
- Labels for the evaporation are not linked up well - Evaporation & Replace NA with- I suppose you know this.
 - The evaporation receiver is not getting focus when variable radio button is checked.
-  When end_of rains checkbox is checked and end of rains column name is empty ok should be disabled. and also, when Replace NA with input is empty?

**Comment by shadrackkibet** [comment]
Testing this again.

**Comment by shadrackkibet** [comment]
- Third point is not yet fixed fully.
image

- When i am on end season, on evaporation variable receiver i have my evaporation column then click ok all works fine. I go back to value radio button i change my value to say 10 then click ok, i get a bug. `evap_NA_as_value` calculation is not properly updated.
image

- It will be good if the default focus receiver on defaults is changed to Rain Column receiver. currently when it resets to the  Date receiver.

**Comment by shadrackkibet** [comment]
Most of the issues here are now fixed. However, this still needs further testing by @rdstern.
-We haven't worked on Getting evaporation column from a linked dataframe ,i suggest we work on this for 0.4.19.

This is ready  for **0.4.18**.

**Comment by rdstern** [comment]
That would be good to test in Malawi.  Once you add the facility to have a linked data frame (for Version 4.19) then one example will be a data frame with 366 rows.  This gives (average) daily values for any year.

**Comment by maxwellfundi** [comment]
@bethanclarke  Are you able to use git bash? You can get rid of the two files. Else a quick fix is to copy paste everything say onto a text file then update your master create a new branch then paste the changes back appropriately. That saves you the agony of using git bash, its quick and easy to do... if thie needs to be in the installer then those changes need to be made soon

**Comment by dannyparsons** [comment]
I think someone manually copying the EOR files into a new branch would be quickest at this point. There are also changes to an R file which are not wanted I think.

**Comment by bethanclarke** [comment]
Sorry about these unwanted file changes here, I thought I have excluded these changes. I have made changed to the designer as well, so what does that mean for copying the files onto another branch?

**Comment by dannyparsons** [comment]
You can copy the designer file code and paste it into the designer file in the same way.

**Pull request closed without merging**